File: ProgrammingBasics/30-Lenses-master/JavaScript/1-lens.js
"use strict";

// Линзы - в функциональном программировании аналог геттеров и сеттеров.
// Первый вызов передает проперти которая будет храниться в замыкании, а второй вызов рабочий который
// выполняет свой функционал.
const getter = (prop) => (obj) => obj[prop];
const setter = (prop) => (val, obj) => ({ ...obj, [prop]: val }); // Сеттер возвращает новый объект в который скопированы
                                                                  // все поля старого и добавлено новое.

const lens = (getter, setter) => ({ // Линза которая на вход принимает функции которые возвращает геттер и сеттер.
  get: (obj) => getter(obj),
  set: (val, obj) => setter(val, obj),
});

// Usage

const nameLens = lens(getter("name"), setter("name"));

console.dir({ nameLens });

File: ProgrammingBasics/30-Lenses-master/Haskell/1-lens.hs
data Person = Person
  { _name :: String
  , _city :: String
  , _born :: Int
  } deriving Show

data Lens s a = Lens (s -> a) (a -> s -> s)
nameGetter = _name
nameSetter name person = person { _name = name }
nameLens = Lens nameGetter nameSetter
